tower-grpc-interop/src/client.rs

```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                        response_stream.into_inner()
+                            .collect()
+                            .map_err(|tower_err: tower_grpc::Error<()>| -> Box<Error> {
+                                Box::new(tower_err)
+                            })
+                    })
+                    .map(|responses: Vec<pb::StreamingOutputCallResponse>| -> Vec<TestAssertion> {
+                        vec![
+                            test_assert!(
+                                "there should be four responses",
+                                responses.len() == 4,
+                                format!("responses.len()={:?}", responses.len())
+                            ),
+                            test_assert!(
+                                "the response payload sizes should match input",
+                                RESPONSE_LENGTHS == response_lengths(&responses).as_slice(),
+                                format!("{:?}={:?}", RESPONSE_LENGTHS, response_lengths(&responses))
+                            ),
+                        ]
+                    })
+                    .then(&assert_success))
+            },
+            Testcase::ping_pong => {
+                fn make_req(idx: usize) -> pb::StreamingOutputCallRequest {
+                    let req_len = REQUEST_LENGTHS[idx];
+                    let resp_len = RESPONSE_LENGTHS[idx];
+                    pb::StreamingOutputCallRequest {
+                        response_parameters: vec![
+                            pb::ResponseParameters::with_size(resp_len)
+                        ],
+                        payload: Some(util::client_payload(req_len as usize)),
+                        ..Default::default()
+                    }
+                };
+
+                /// Contains the state that is threaded through the recursive "loop"
+                /// of futures that performs the actual ping-pong with the server.
+                struct State {
+                    sender: futures::sync::mpsc::UnboundedSender<pb::StreamingOutputCallRequest>,
+                    stream: Box<Stream<Item=pb::StreamingOutputCallResponse, Error=tower_grpc::Error>>,
+                    responses: Vec<pb::StreamingOutputCallResponse>,
+                    assertions: Vec<TestAssertion>,
+                };
+
+                type ResponsesFuture = Box<Future<
+                    Item=(Vec<pb::StreamingOutputCallResponse>, Vec<TestAssertion>),
+                    Error=Box<Error>>>;
+                /// Recursive function that takes the current State of the ping-pong
+                /// operation and performs the rest of it. It returns a future of
+                /// responses from the server and assertions that were performed in
+                /// the process.
+                fn perform_ping_pong(state: State) -> ResponsesFuture {
+                    let (sender, stream, mut responses, mut assertions) =
```

Comment:
Nit: I think this might be somewhat nicer as
```rust
let State { sender, stream, mut responses, mut assertions } = state;
```

Reply:
Nice! I don't know why I didn't do that in the first place. Fixed.